### PR TITLE
Add advice from resources

### PR DESCRIPTION
**Pull request by ShaneDelmore**
This pr updates clippy to collect advice from all clippy.json files in resources.
It also returns all matching advice, that way if multiple library authors register help for the same message (say .toJson not found) then we can display all of their advice and allow the user to choose.

### CODE REVIEW

**Comment by ShaneDelmore** [comment]
If this is acceptable then once it is published I will make PRs to a few libraries I use and see if I can get them to pull in the advice.  Let me know if you have any suggested libraries to start with.

**Comment by ShaneDelmore** [comment]
CompileTests.scala work for me locally running sbt test, but fail on Travis.  Not sure why yet.

**Comment by ShaneDelmore** [comment]
It looks like the master branch is also failing on travis.  I am rebasing my changes back onto release 3.1 which looks like the last successful build on Travis.

**Comment by ShaneDelmore** [comment]
I have verified I get a gree build when reverting the 2.12 clippy changes.
https://travis-ci.org/softwaremill/scala-clippy/builds/175406394

**Comment by adamw** [comment]
Thanks! First, let me look at the failing 2.12 build :)

**Comment by ShaneDelmore** [comment]
It was ScalaTest causing the failure. I wasn't able to narrow it down beyond that.

**Comment by adamw** [comment]
Yeah, but locally everything works, which makes debugging a bit hard. And we do want 2.12 compatibility :) I'm trying with a newer version of scalatest now.

**Comment by adamw** [comment]
Ok, travis build fixed, there was an old scala version used in the `.travis.yml` file. Could you rebase your changes to master?

**Comment by ShaneDelmore** [comment]
No problem, I can do it in an hour or two.

File: ui-client/src/main/scala/com/softwaremill/clippy/AutowireClient.scala [hunk 1]
```diff
@@ -2,7 +2,7 @@ package com.softwaremill.clippy
 
 import org.scalajs.dom
 import scala.concurrent.Future
-import scalajs.concurrent.JSExecutionContext.Implicits.runNow
+import scalajs.concurrent.JSExecutionContext.Implicits.queue
```
Comment (adamw) [inline ui-client/src/main/scala/com/softwaremill/clippy/AutowireClient.scala]: is that change needed as well?

File: tests/src/test/scala/org/softwaremill/clippy/CompileTests.scala [hunk 2]
```diff
@@ -110,24 +112,26 @@ class CompileTests extends FlatSpec with Matchers with BeforeAndAfterAll {
...
-      catch {
-        case e: ToolBoxError =>
-          e.message should include("Clippy advises")
-      }
+      (the[ToolBoxError] thrownBy parse(s)).message should include("Clippy advises")
     }
   }
+
+  "Clippy" should "return all matching advice" in {
+    (the[ToolBoxError] thrownBy parse(snippets("macwire")))
```
Comment (adamw) [inline tests/src/test/scala/org/softwaremill/clippy/CompileTests.scala]: nice, reads much better :)

File: plugin/src/main/scala/com/softwaremill/clippy/AdviceLoader.scala [hunk 3]
```diff
@@ -10,12 +10,25 @@ import scala.concurrent.{ExecutionContext, Future}
...
+    getClass.getClassLoader
+      .getResources("clippy.json")
+      .toIterator
+      .map{ url =>
+        val is = url.openStream()
+        val clips = inputStreamToClippy(is)
+        is.close()
+        clips.advices
+      }
+  }.flatten.toList
```
Comment (adamw) [inline plugin/src/main/scala/com/softwaremill/clippy/AdviceLoader.scala]: I would add some error-handling here. E.g. wrap reading advice from each URL in a `Try`, and then collect only the successful ones, discarding (logging a warning) the ones that failed. That way if a single library would contain malform `clippy.json` file, that wouldn't break the whole compilation process.

Comment (ShaneDelmore) [inline plugin/src/main/scala/com/softwaremill/clippy/AdviceLoader.scala]: Great idea, thanks.

**Comment by ShaneDelmore** [comment]
I will make an update in a day or two about adding advice in clippy.json.  I will also reach out to a few library authors once it is published and see if they are receptive to me adding clippy advice and tests to their repos.

**Comment by ShaneDelmore** [comment]
I added minimal information to the readme on including library specific advice.  Once it is published and I can verify the concept works well for my internal libraries I would like to extend the web interface to show the json produced and reference that in the readme as a way for library authors to easily create the Advice they need.

**Comment by ShaneDelmore** [comment]
Updated to implement feedback from @adamw

**Comment by adamw** [comment]
0.3.2 released - should be available shortly

**Comment by ShaneDelmore** [comment]
0.3.2 compiler plugin is available but no sbt plugin.  I don't actually use the sbt-plugin anymore, just double checked since it failed to deploy last time.

**Comment by adamw** [comment]
Ah thanks! It got published this time, but for Scala 2.11, which isn't used by SBT ;) Fixing this now
